### ISSUE TITLE
On NodeLost, the new pod can't mount the same volume.

### ISSUE THREAD

**Issue by dimm0** [rook/rook#1507]
Is this a bug report or feature request?
* Bug Report

**Bug Report**

What happened:
A PV was attached to a POD. The node running the POD lost the power. Currently the status for the node is NotReady.

Deployment is trying to restart the POD on another node, but rook doesn't let the volume remount:

```
  Warning  FailedMount  16m (x82 over 2h)  kubelet, fiona-dtn-1.ucsc.edu  MountVolume.SetUp failed for volume "pvc-31b7555e-122d-11e8-9422-0cc47a6be994" : mount command failed, status: Failure, reason: Rook: Mount volume failed: failed to attach volume pvc-31b7555e-122d-11e8-9422-0cc47a6be994 for pod perfsonar/perfsonar-toolkit-bdc95f48f-xhzt7. Volume is already attached by pod perfsonar/perfsonar-toolkit-bdc95f48f-zlhld. Status Running
  Warning  FailedMount  2m (x73 over 2h)   kubelet, fiona-dtn-1.ucsc.edu  Unable to mount volumes for pod "perfsonar-toolkit-bdc95f48f-xhzt7_perfsonar(2394445f-138c-11e8-9422-0cc47a6be994)": timeout expired waiting for volumes to attach/mount for pod "perfsonar"/"perfsonar-toolkit-bdc95f48f-xhzt7". list of unattached/unmounted volumes=[pgsql-persistent-storage]
```

The status for old pod is not running, as it's claiming:
```
Dmitrys-MBP-2:~ dimm$ kubectl describe pod perfsonar-toolkit-bdc95f48f-zlhld -n perfsonar
Name:                      perfsonar-toolkit-bdc95f48f-zlhld
Namespace:                 perfsonar
Node:                      k8s-chase-ci-06.calit2.optiputer.net/67.58.53.163
Start Time:                Thu, 15 Feb 2018 00:49:55 -0800
Labels:                    k8s-app=perfsonar-toolkit
                           pod-template-hash=687519049
Annotations:               kubernetes.io/limit-ranger=LimitRanger plugin set: memory request for container perfsonar-toolkit; memory limit for container perfsonar-toolkit; memory request for init container volume-mount-chown; m...
Status:                    Terminating (expires Fri, 16 Feb 2018 18:42:34 -0800)
Termination Grace Period:  30s
Reason:                    NodeLost
Message:                   Node k8s-chase-ci-06.calit2.optiputer.net which was running pod perfsonar-toolkit-bdc95f48f-zlhld is unresponsive
IP:                        10.244.5.6
Controlled By:             ReplicaSet/perfsonar-toolkit-bdc95f48f
```

What you expected to happen:
When node is lost, a pod having PV mounted should restart somewhere else without human intervention.

How to reproduce it (minimal and precise):
Start a deployment with a pod mounting rook PV. Kill the physical node running the pod.

**Environment**:
* OS (e.g. from /etc/os-release):
CentOS Linux 7 (Core)
* Kernel (e.g. `uname -a`):
Linux 4.14.15-1.el7.elrepo.x86_64 #1 SMP Tue Jan 23 20:28:26 EST 2018 x86_64 x86_64 x86_64 GNU/Linux
* Cloud provider or hardware configuration:
Baremetal
* Rook version (use `rook version` inside of a Rook Pod):
rook: v0.6.0-219.g09dc9b8
* Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T10:09:24Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-07T11:55:20Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
kubeadm

**Comment by karstensi** [comment]
I have the exact same experience on environment:
* OS
  Debian GNU/Linux 9 (stretch)
* Kernel
  Linux <hostname> 4.9.0-5-amd64 #1 SMP Debian 4.9.65-3+deb9u2 (2018-01-04) x86_64 GNU/Linux
* Cloud provider
  AWS
* Rook version
 rook: v0.7.0-4.g8064b28
* kubernets version:
  Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-07T12:22:21Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-07T11:55:20Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
* Kubernetes cluster type
  kubeadm 1.9.3

**Comment by dimm0** [comment]
Yup, never goes away until the node is started or deleted from the cluster. K8s can't clear the pod from non-running node..

**Comment by travisn** [comment]
ok, looks like we will need to add that condition for attaching.

**Comment by galexrt** [comment]
@travisn The problem I see is that `Terminating` doesn't have to mean the Pod is already down on the broken node it still could be running and still consuming the storage. For example it could possibly be that the node the Pod runs on just has an issue with kubelet and kubelet has "crashed"/been stopped for some reason where the Pods continue running.

**Comment by lorenz** [comment]
As far as I know this needs RBD fencing which is not properly supportable in container-based environments without full-force unmapping. Basically one needs to fence the old node of the volume using Ceph fencing before reattaching to a different node (to prevent corruption arising from two nodes having the device mapped). The old node needs to execute a full-force unmap (discarding all IO) once it resynchronizes and terminates the pod. This is not supportet ATM, see https://github.com/rook/rook/issues/376#issuecomment-317172298

**Comment by PaulCapestany** [comment]
FWIW I seem to be running into this issue as well. Even **_after_** restarting the previously-powered-off node the logs still show `Volume is already attached by pod`, and this prevents a new pod from attaching (e.g., having followed Rook's Wordpress tutorial, the `wordpress-mysql` pod ends up stuck in a `ContainerCreating` status).

**Comment by guilhermeblanco** [comment]
Experienced the same issue today during `kops` upgrade from 1.8.7 to 1.9.6.
I have a dedicated `InstanceGroup` of i3 nodes called `nodes-data`. During the upgrade, the instances are cordoned and new nodes are brought up. This means that NVMe nodes-data will have their data trashed.
As the upgrade progressed, upon first k8s worker node, I started to face FailedMount issues, similar to the previous comment. `ceph health` pointed the HEALTH_ERROR, and it was marking as 3 OSDs down. I tried to recover by manually tagging ceph that OSDs were lost, but the cluster refused to take any action. It also mentioned that some PGs were staled, and that I have a 36% degraded PGs.
Another attempt taken was to manually remove the `VolumeAttachment` from evicted pods, without success.

Our team agreed to accept this issue as a catastrophic disaster and accept the data loss. Luckily enough, this happened to our development k8s cluster (we have one per environment and region).

Even though I mentioned I had no more console outputs of what I experienced, I was asked to provide my scenario here.

**Comment by lorenz** [comment]
@guilhermeblanco This issue is about mounting volumes previously attached to pods without clean detachment (because the node hung/disconnected/crashed). Your issue is about your underlying Ceph cluster failing. Just a small comment about that: Ceph will not gracefully handle trashing its data. The cluster will fail and usually take your data with it. If you need to migrate the data, you need to explicitly tell Ceph to migrate the data (and the mons) to the new nodes and wait for that process to complete before nuking the old ones.

**Comment by galexrt** [comment]
@dimm0 @guilhermeblanco @lorenz Is still happening with the `0.7.1` release?
At least for me it seems that this issue isn't occurring anymore.

**Comment by lorenz** [comment]
@galexrt This definitely happens in 0.7.1. Spawn in 3 nodes and blow one up by either purposely crashing it (echo c > /proc/sysrq-trigger) or fully disconnecting all networking. Then a new pod will be spawned which will stay in Creating until the original node is brought back up again.

**Comment by motecshine** [comment]
`
  Normal   Scheduled              19s               default-scheduler  Successfully assigned ebills-694dbc577d-6t8fm to node2
  Normal   SuccessfulMountVolume  18s               kubelet, node2     MountVolume.SetUp succeeded for volume "log"
  Normal   SuccessfulMountVolume  18s               kubelet, node2     MountVolume.SetUp succeeded for volume "code"
  Normal   SuccessfulMountVolume  18s               kubelet, node2     MountVolume.SetUp succeeded for volume "default-token-zwknq"
  Warning  FailedMount            0s (x6 over 18s)  kubelet, node2     MountVolume.SetUp failed for volume "pvc-3993452c-7dde-11e8-b8af-080027f7088b" : mount command failed, status: Failure, reason: Rook: Mount volume failed: failed to attach volume pvc-3993452c-7dde-11e8-b8af-080027f7088b for pod ebills/ebills-694dbc577d-6t8fm. Volume is already attached by pod
`

@galexrt   0.7.1, this problem seems dose not resolved

**Comment by galexrt** [comment]
Can you please try out these settings https://fatalfailure.wordpress.com/2016/06/10/improving-kubernetes-reliability-quicker-detection-of-a-node-down/ and report if it still happens.The link shows how to lower the timeout for node NotReady detection and decrease the time to pod eviction of Pods on NotReady nodes.

**Comment by snowmansora** [comment]
Hi @galexrt ,

I have tried the settings you mentioned in https://fatalfailure.wordpress.com/2016/06/10/improving-kubernetes-reliability-quicker-detection-of-a-node-down/ with just a little change on one of the values:
```
--node-monitor-grace-period=20s (the link suggests 16s)
--node-monitor-period=2s
--pod-eviction-timeout=30s
--node-status-update-frequency=4s 
```

I am using rook ceph block v0.8.0, and I still see the problem reported in this issue.

Below contains the details of my setup.

- OS: CentOS Linux release 7.5.1804 (Core)
- Kernel: 3.10.0-862.3.2.el7.x86_64
- Cloud provider: AWS 3 m4.large EC2 instances
- Rook version: v0.8.0
- Kubernetes version:
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.5", GitCommit:"f01a2bf98249a4db383560443a59bed0c13575df", GitTreeState:"clean", BuildDate:"2018-03-19T15:50:45Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.5", GitCommit:"f01a2bf98249a4db383560443a59bed0c13575df", GitTreeState:"clean", BuildDate:"2018-03-19T15:50:45Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
- Kubernetes cluster type: Kubespray with flannel CNI

Deployed rook ceph block via applying the following yaml files:
1. https://github.com/rook/rook/blob/release-0.8/cluster/examples/kubernetes/ceph/operator.yaml (pinned docker image to v0.8.0 and updated FLEXVOLUME_DIR_PATH)
2. https://github.com/rook/rook/blob/release-0.8/cluster/examples/kubernetes/ceph/cluster.yaml
3. https://github.com/rook/rook/blob/release-0.8/cluster/examples/kubernetes/ceph/storageclass.yaml

<details><summary>K8s Deployment and StatefulSet to reproduce the problem</summary>
<p>

```
---
kind: Service
apiVersion: v1
metadata:
  name: rook-test
  labels:
    app: rook-test
spec:
  type: ClusterIP
  ports:
  - name: rook-test
    port: 80
  selector:
    app: rook-test
---
kind: StatefulSet
apiVersion: apps/v1beta1
metadata:
  name: rook-test
spec:
  serviceName: rook-test
  replicas: 1
  volumeClaimTemplates:
  - metadata:
      name: data
    spec:
      storageClassName: rook-ceph-block
      accessModes:
        - "ReadWriteOnce"
      resources:
        requests:
          storage: 10Gi
  template:
    metadata:
      labels:
        app: rook-test
    spec:
      containers:
      - name: rook-test
        image: nginxdemos/hello
        ports:
        - containerPort: 80
        volumeMounts:
        - name: data
          mountPath: /rook
---
kind: Service
apiVersion: v1
metadata:
  name: rook-test2
  labels:
    app: rook-test2
spec:
  type: ClusterIP
  ports:
  - name: rook-test2
    port: 80
  selector:
    app: rook-test2
---
kind: Deployment
apiVersion: apps/v1
metadata:
  name: rook-test2
spec:
  selector:
    matchLabels:
      app: rook-test2
  replicas: 1
  template:
    metadata:
      labels:
        app: rook-test2
    spec:
      containers:
      - name: rook-test2
        image: nginxdemos/hello
        ports:
        - containerPort: 80
```

</p>
</details>

In my 3 nodes cluster, I have shut down k8s-node-002, and it became status NotReady:
```
[centos@k8s-node-001 ~]$ kubectl get nodes -o wide
NAME           STATUS     ROLES                 AGE       VERSION   EXTERNAL-IP   OS-IMAGE                KERNEL-VERSION              CONTAINER-RUNTIME
k8s-node-001   Ready      ingress,master,node   53m       v1.9.5    <none>        CentOS Linux 7 (Core)   3.10.0-862.3.2.el7.x86_64   docker://17.3.2
k8s-node-002   NotReady   ingress,master,node   53m       v1.9.5    <none>        CentOS Linux 7 (Core)   3.10.0-862.3.2.el7.x86_64   docker://17.3.2
k8s-node-003   Ready      ingress,master,node   53m       v1.9.5    <none>        CentOS Linux 7 (Core)   3.10.0-862.3.2.el7.x86_64   docker://17.3.2
```

rook-test-0 (StatefulSet) which uses rook ceph block stuck in status Unknown while rook-test2- (Deployment) which does not use rook ceph block failed-over to another node quickly. p.s. I have waited 10 minutes+ after the node shutdown:
```
[centos@k8s-node-001 ~]$ kubectl get pods -o wide
NAME                          READY     STATUS    RESTARTS   AGE       IP             NODE
rook-test-0                   1/1       Unknown   0          46m       10.233.64.13   k8s-node-002
rook-test2-5fd4f659cc-678gg   1/1       Running   0          38m       10.233.65.14   k8s-node-003
rook-test2-5fd4f659cc-r2gqb   1/1       Unknown   0          46m       10.233.64.12   k8s-node-002
```

Details of rook-test-0 show status Terminating (p.s. not sure why kubectl shows status Unknown though):
```
[centos@k8s-node-001 ~]$ kubectl describe pod rook-test-0
Name:                      rook-test-0
Namespace:                 default
Node:                      k8s-node-002/10.10.3.173
Start Time:                Mon, 30 Jul 2018 08:30:06 +0000
Labels:                    app=rook-test
                           controller-revision-hash=rook-test-7b8464c9d8
                           statefulset.kubernetes.io/pod-name=rook-test-0
Annotations:               <none>
Status:                    Terminating (expires Mon, 30 Jul 2018 08:38:48 +0000)
Termination Grace Period:  30s
Reason:                    NodeLost
Message:                   Node k8s-node-002 which was running pod rook-test-0 is unresponsive
IP:                        10.233.64.13
Controlled By:             StatefulSet/rook-test
Containers:
  rook-test:
    Container ID:   docker://cdcdcd8cb1e5a949e6a2017b43c73a19dc059ce2d3b81842ccbcce1346f7ca5c
    Image:          nginxdemos/hello
    Image ID:       docker-pullable://nginxdemos/hello@sha256:f5a0b2a5fe9af497c4a7c186ef6412bb91ff19d39d6ac24a4997eaed2b0bb334
    Port:           80/TCP
    State:          Running
      Started:      Mon, 30 Jul 2018 08:30:12 +0000
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /rook from data (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-lgzpq (ro)
Conditions:
  Type           Status
  Initialized    True
  Ready          False
  PodScheduled   True
Volumes:
  data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  data-rook-test-0
    ReadOnly:   false
  default-token-lgzpq:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-lgzpq
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     <none>
Events:
  Type    Reason                  Age   From                   Message
  ----    ------                  ----  ----                   -------
  Normal  Scheduled               53m   default-scheduler      Successfully assigned rook-test-0 to k8s-node-002
  Normal  SuccessfulMountVolume   53m   kubelet, k8s-node-002  MountVolume.SetUp succeeded for volume "default-token-lgzpq"
  Normal  SuccessfulMountVolume   53m   kubelet, k8s-node-002  MountVolume.SetUp succeeded for volume "pvc-c32b05ac-93d2-11e8-bee6-12dfceee12ee"
  Normal  Pulling                 53m   kubelet, k8s-node-002  pulling image "nginxdemos/hello"
  Normal  Pulled                  53m   kubelet, k8s-node-002  Successfully pulled image "nginxdemos/hello"
  Normal  Created                 53m   kubelet, k8s-node-002  Created container
  Normal  Started                 53m   kubelet, k8s-node-002  Started container
  Normal  NodeControllerEviction  45m   node-controller        Marking for deletion Pod rook-test-0 from Node k8s-node-002
```

I also tried to delete the pod forcefully:
```
kubectl delete pod rook-test-0 --grace-period=0 --force
```

Then, rook-test-0 is rescheduled to another node but stuck at status ContainerCreating:
```
[centos@k8s-node-001 ~]$ kubectl get pods -o wide
NAME                          READY     STATUS              RESTARTS   AGE       IP             NODE
rook-test-0                   0/1       ContainerCreating   0          2m        <none>         k8s-node-003
rook-test2-5fd4f659cc-678gg   1/1       Running             0          52m       10.233.65.14   k8s-node-003
rook-test2-5fd4f659cc-r2gqb   1/1       Unknown             0          1h        10.233.64.12   k8s-node-002
```

Details of rook-test-0 shows timeout expired waiting for volumes to attach/mount for pod:
```
[centos@k8s-node-001 ~]$ kubectl describe pod rook-test-0
Name:           rook-test-0
Namespace:      default
Node:           k8s-node-003/10.10.3.207
Start Time:     Mon, 30 Jul 2018 09:28:39 +0000
Labels:         app=rook-test
                controller-revision-hash=rook-test-7b8464c9d8
                statefulset.kubernetes.io/pod-name=rook-test-0
Annotations:    <none>
Status:         Pending
IP:
Controlled By:  StatefulSet/rook-test
Containers:
  rook-test:
    Container ID:
    Image:          nginxdemos/hello
    Image ID:
    Port:           80/TCP
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /rook from data (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-lgzpq (ro)
Conditions:
  Type           Status
  Initialized    True
  Ready          False
  PodScheduled   True
Volumes:
  data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  data-rook-test-0
    ReadOnly:   false
  default-token-lgzpq:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-lgzpq
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     <none>
Events:
  Type     Reason                 Age   From                   Message
  ----     ------                 ----  ----                   -------
  Normal   Scheduled              2m    default-scheduler      Successfully assigned rook-test-0 to k8s-node-003
  Normal   SuccessfulMountVolume  2m    kubelet, k8s-node-003  MountVolume.SetUp succeeded for volume "default-token-lgzpq"
  Warning  FailedMount            27s   kubelet, k8s-node-003  Unable to mount volumes for pod "rook-test-0_default(f12dee76-93da-11e8-bee6-12dfceee12ee)": timeout expired waiting for volumes to attach/mount for pod "default"/"rook-test-0". list of unattached/unmounted volumes=[data]

```

Log of rook-ceph-agent on the node where new rook-test-0 tries to spawn:
```
[centos@k8s-node-001 ~]$ kubectl log rook-ceph-agent-7cswj -n rook-ceph-system
W0730 09:36:02.753812   28017 cmd.go:353] log is DEPRECATED and will be removed in a future version. Use logs instead.
2018-07-30 08:23:41.122886 I | rookcmd: starting Rook v0.8.0 with arguments '/usr/local/bin/rook ceph agent'
2018-07-30 08:23:41.122952 I | rookcmd: flag values: --help=false, --log-level=INFO
2018-07-30 08:23:41.124483 I | cephcmd: starting rook ceph agent
2018-07-30 08:23:41.124513 I | exec: Running command: modinfo -F parm rbd
2018-07-30 08:23:41.130505 I | exec: Running command: modprobe rbd single_major=Y
2018-07-30 08:23:41.243484 I | flexvolume: Listening on unix socket for Kubernetes volume attach commands: /flexmnt/ceph.rook.io~rook-ceph-system/.rook.sock
2018-07-30 08:23:41.315918 I | flexvolume: Listening on unix socket for Kubernetes volume attach commands: /flexmnt/ceph.rook.io~rook/.rook.sock
2018-07-30 08:23:41.388183 I | flexvolume: Listening on unix socket for Kubernetes volume attach commands: /flexmnt/rook.io~rook-ceph-system/.rook.sock
2018-07-30 08:23:41.673411 I | flexvolume: Listening on unix socket for Kubernetes volume attach commands: /flexmnt/rook.io~rook/.rook.sock
2018-07-30 08:23:41.676013 I | agent-cluster: start watching cluster resources
E0730 08:37:48.313119       5 streamwatcher.go:109] Unable to decode an event from the watch stream: http2: server sent GOAWAY and closed the connection; LastStreamID=15, ErrCode=NO_ERROR, debug=""
2018-07-30 09:28:39.760748 I | flexdriver: calling agent to attach volume replicapool/pvc-c32b05ac-93d2-11e8-bee6-12dfceee12ee
2018-07-30 09:28:39.764956 I | flexvolume: Volume attachment record rook-ceph-system/pvc-c32b05ac-93d2-11e8-bee6-12dfceee12ee exists for pod: default/rook-test-0
2018-07-30 09:28:39.767873 I | flexvolume: Volume attachment record rook-ceph-system/pvc-c32b05ac-93d2-11e8-bee6-12dfceee12ee is orphaned. Updating record with new attachment information for pod default/rook-test-0
2018-07-30 09:28:39.785471 I | rook-ceph-agent-crd: updated Volumeattach CRD pvc-c32b05ac-93d2-11e8-bee6-12dfceee12ee
2018-07-30 09:28:39.785552 I | ceph-volumeattacher: attaching volume replicapool/pvc-c32b05ac-93d2-11e8-bee6-12dfceee12ee cluster rook-ceph
2018-07-30 09:28:39.792011 I | cephmon: parsing mon endpoints: rook-ceph-mon2=10.233.50.122:6790,rook-ceph-mon3=10.233.32.133:6790,rook-ceph-mon0=10.233.54.72:6790
2018-07-30 09:28:39.792102 I | op-mon: loaded: maxMonID=3, mons=map[rook-ceph-mon2:0xc42042f720 rook-ceph-mon3:0xc42042f7a0 rook-ceph-mon0:0xc42042f840], mapping=&{Node:map[rook-ceph-mon0:0xc4202e5680 rook-ceph-mon2:0xc4202e5770 rook-ceph-mon3:0xc4202e57d0] Port:map[]}
2018-07-30 09:28:39.792349 I | exec: Running command: rbd map replicapool/pvc-c32b05ac-93d2-11e8-bee6-12dfceee12ee --id admin --cluster=rook-ceph --keyring=/tmp/rook-ceph.keyring659539085 -m 10.233.50.122:6790,10.233.32.133:6790,10.233.54.72:6790 --conf=/dev/null
```

If you need more information, I will try my best to provide.

I think this bug is really a blocker for fail over use case.

Thank you for your help in advance!

**Comment by gokulpch** [comment]
@galexrt @snowmansora I am seeing this issue on Kubernetes-1.11.1-0 with replicapool replication set to 3 and  not using StatefulSet. The functionality (mapping/re-mapping of volumes) works without issues if a Cordon and Drain is performed on the node, but the same fails in the case of reboot and the node pauses while booting with the message below:

<img width="822" alt="screen shot 2018-08-05 at 1 03 22 am" src="https://user-images.githubusercontent.com/18540565/43736735-8b4f525e-9973-11e8-92a6-8bc655e294ad.png">

The node requires a power-off and on after this. I tweaked all the parameters as mentioned to quickly detect the nodelost/node-failure and Kubernetes tries to schedule the pod on the other node but fails with the message below:

```
Events:
  Type     Reason       Age               From               Message
  ----     ------       ----              ----               -------
  Normal   Scheduled    1m                default-scheduler  Successfully assigned default/prometheus-server-796945cbb9-njlkh to node2
  Warning  FailedMount  38s (x8 over 1m)  kubelet, node2     MountVolume.SetUp failed for volume "pvc-b8b02e65-9882-11e8-a9fc-525400a4cd50" : mount command failed, status: Failure, reason: Rook: Mount volume failed: failed to attach volume pvc-b8b02e65-9882-11e8-a9fc-525400a4cd50 for pod default/prometheus-server-796945cbb9-njlkh. Volume is already attached by pod default/prometheus-server-796945cbb9-r5txb. Status Running
```

**Comment by davidhiendl** [comment]
Is there any way to manually mark a volume as detached (if it is not possible automatically at the moment)? This is a real problem for applications that do not support HA directly.

**Comment by lorenz** [comment]
@davidhiendl You can delete the VolumeAttachment Resource but that is very dangerous since there is no fencing on the corresponding RBD.

**Comment by davidhiendl** [comment]
@lorenz Thanks for the info.

So basically there is currently no way with a Ceph cluster managed by Rook to make a single-instance application failover to an other node when using rbd, right? This is a huge problem for us and for many other people I think. Unfortunately some legacy applications simply don't scale horizontally or support failover ...

As far as I understand it this problem does not exist when deploying Ceph directly on a host (we were hoping to avoid it but if it is not possible then we might have to)?

**Comment by lorenz** [comment]
@davidhiendl I totally agree with you that this is a fairly substantial issue (for the same reasons you mentioned). I'm going to write my thoughts down on how I think this could work (take this with a grain of salt, I have a fair bit of experience with distributed systems but I'm not part of Rook or Ceph).

The following cases need to be considered:
1. Node blows up (Triple-fault, Kernel Panic, CPU goes up in smoke, Power goes off, CPU hangs itself indefinitely, ...)
2. Node hangs for a long but definite amount of time
3. Node gets partitioned (routing issue, cable got pulled, ...)

In all cases we want to reattach the RBD volume to another node while retaining all data and simultaneously preventing the previous node from reading or writing from the dataset.
Case 1 is fairly trivial to handle, just reattach and remount somewhere else. Cases 2 and 3 require a lock on the RBD so that only one node can have it attached at the same time thus preventing writes from stale nodes (this is called fencing) because otherwise the process on the faulty node can just queue up writes and will eventually issue all of them as soon as it starts working again (which will corrupt data). Ceph has two locking systems, the old one and the new one (Exclusive Locking). The new one should be able to support this usecase, but there exists conflicting information if its lock transfers are cooperative or not (if they are this would mean that that locking mechanism would be unsuitable for HA purposes).

The problem here is that as soon as you use RBD fencing the machine being fenced cannot issue any IO requests to the cluster and all of them remain pending. There is currently no way to unmap an RBD device with pending IO requests (https://github.com/ceph/ceph-client/blob/8b8f53af1ed9df88a4c0fbfdf3db58f62060edf3/drivers/block/rbd.c#L5857) so you are left with unkillable processes and a dead device which bloats your System Load and a possible deathtrap for any management process that touches that device because it will immediately go into Unkillable IO sleep state.

Basically three things need to happen to make Rook work for your usecase:
1. Rook needs to use Ceph's exclusive locking and not its own since it cannot implement suitable fencing due to atomicity constraints.
2. Ceph RBD needs to implement a force-unmap option to discard all pending IO so that a machine which has been fenced is not a disaster waiting to happen
3. Rook needs to watch the RBD locks and kill all associated resources (containers, mounts, maps) if it detects that the node lost its exclusive RBD lock for a volume

**Comment by jbw976** [comment]
@travisn @rootfs is there anyone at redhat that is an RBD expert that can comment on any possible updates in this scenario.  To summarize, if contact with a node in the cluster is lost, is there any way for another node in the cluster to confidently know that it is safe to start up a new client to that same RBD volume?  Can the new node "forcefully" remove the old client?

**Comment by jbw976** [comment]
Adding to 0.9 since this is hurting production deployments such as @dimm0

**Comment by dac88** [comment]
From a users perspective i think the only option is to apply a node selector for a single node and attempt to duplicate the entire setup for resilience. Not ideal by any means as data is free to become out of sync.

**Comment by vistalba** [comment]
I run into the same issue with rook. :(

One Node of a 3 node cluster goes down. Kubernetes tries to restart the pods on another node but they stuck in state "ContainerCreating" because the volume is already attached to the pod in state "Terminating".

Is anyone working on this issue? Is there any workaround or something like this?
I think this is a very important scenario and should be possible.

**Comment by lorenz** [comment]
@vistalba Look at https://github.com/rook/rook/issues/1507#issuecomment-413604790, I've summarized the current state and possible solutions there.

**Comment by vistalba** [comment]
@lorenz Yes I saw this post but it's old and unfortunatly it looks like there is no progress :(

**Comment by davidhiendl** [comment]
I was wondering if this (or similar) issue does apply to EdgeFS? I've read somewhere it already supports fencing and therefor might a viable workaround/alternative.

**Comment by vistalba** [comment]
Can someone confirm that the EdgeFS would work for this scenario mentioned by @davidhiendl ?

**Comment by lorenz** [comment]
At best partially (as in the new node can get the volume). But at least the equivalent of a force_unmap option is not available on either iSCSI or NFS which are used as access protocols by EdgeFS.

**Comment by dyusupov** [comment]
> Can someone confirm that the EdgeFS would work for this scenario mentioned by @davidhiendl ?

EdgeFS allows multi-head NFS exports with full RW semantics. See "instances" CRD parameter. However, please be advised that application-level synchronization still required, i.e. use of fcntl(2).

**Comment by dimm0** [comment]
NFS is slow in general. I'd be curious how it works with iSCSI and ReadOnce PV

**Comment by dimm0** [comment]
I'm talking about this specific issue. Applications using the block device in single-access mode. Those need to failover to a new node when node fails.

**Comment by zerkms** [comment]
Hi,

I'm a novice rook and kubernetes user yet, and while assessing k8s+rook as a platform for my new infrastructure I also experienced this problem while trying various scenarios.

From what @lorenz suggested it looks like Ceph also needs to introduce some changes to RBD, which may take long time or not happen ever.

I understand that what I'm about to suggest is naive, but I believe being optional it may save a lot of problems while waiting for a proper solution. This is not a well thought yet idea, but to my humble opinion it looks like it should work (to some degree :-) ).

Well, what if we had another concept, called `rook-fencer`, an `http` service (for simplicity, so that everybody could implement it any way they want), that is delivered/configured by an administrator specifically for the infrastructure: be it aws, gcp, bare metal, or whatever else.

Its (the `rook-fencer`'s) purpose is to fence the node on demand.

Typical scenario: let's say we run in AWS and one of worker nodes kernel-panics. After some time kubernetes marks that node as unavailable and starts un-scheduling payloads from it. Stateless pods are evicted easily. But the ones with rook-ceph (or other rook-plugins) are blocked, since the rook agent may not unmount it.

Here is where `rook-fencer` comes into play: after a (configurable) timeout a rook agent (the one that serves the node where the new pod is scheduled) performs an http call to the fencer service.

Given that a fencer service is implemented by the administrator - it may do whatever it needs to ensure that the node is properly fenced. In one case it might be a firewall, that blocks all ceph traffic to and from the node. In other cases - it might be node shutdown or restart.

Let's say in this case the administrator implemented a fencer service that shuts down the node. So if a node is in running state - it sends a shutdown request. And polls while it has not shut down. After it is shutdown - it returns `HTTP 200` or any other response that means "the fencing operation completed successfully".

After this point - the agent knows, that the node is properly fenced, no traffic is to be sent from it to ceph, so it's safe to remove the volume mount from the kubernetes api.

This solution allows the implementation to be as simple or as robust as a particular infrastructure requires. For bare metal infrastructures it may be interaction with idrac's, hardware routers or firewalls, etc. For cloud infrastructures - it's even simpler, given they all provide APIs.

Technically `rook` project may provide simple implementations for common cloud providers, but the idea is to design an API that the rook agents may consume.

**Comment by akkie** [comment]
I've the same issue during the update of helm charts. Due to kubernetes rolling update feature, it happens that if the updated version of a pod gets deployed on another node, then it cannot mount the volume, because it's still mounted by the running pod.

**Comment by davidhiendl** [comment]
@akkie That does not sound related to this issue (which is volumes are locked once a server fails). A regular Block Volumes can only be rw mounted once (unlike the CephFS which is a multi rw volume). If you have a pod using a block volume you need to make sure it is stopped first and then start a new one: For example with the deployment strategy like recreate.

**Comment by BlaineEXE** [comment]
We believe this issue will be fixed with CSI, which we intend to be a total replacement for FlexVolume as soon as CSI is ready. CSI will be in beta status for the v1.0 release targeted in the next few weeks, and CSI features are still being finalized. 

This issue was just moved to the v1.1 release milestone, as we hope CSI will be at a stable release by then. As part of the CSI stable release, we intend to follow the original issue's repro steps to ensure that the issue does not exist in the CSI release implementation.

**Comment by danielxolsson** [comment]
Solved this by using following config:
     terminationGracePeriodSeconds: 0
     tolerations:
     - effect: NoExecute
       key: node.kubernetes.io/unreachable
       operator: Exists
       tolerationSeconds: 300

Tested with kubernetes v1.12 with rook-ceph v14.2. Note that FlexVolume is used.

**Comment by dimm0** [comment]
How does it prevent node from reconnecting to ceph and damaging the ceph volume when network conenction is reestablished?

**Comment by danielxolsson** [comment]
Below some assumptions:

If node is just shutdown and when booted it will reconnect with kubernetes, then the pod will not start at all. 

In a split-brain scenario ceph will not allow a second rw mount of the volume. The opposite is true if node lost connection to kubernetes, but has connection to ceph. 

Just wanted to share what worked for me...

**Comment by zerkms** [comment]
> ceph will not allow a second rw mount of the volume

is it actually the case?

**Comment by davidhiendl** [comment]
What about writes inside a network partition that occurred in the timeframe where the node was partitioned and not yet evicted? If the partition reconnects and the main group has had writes since then there is no way to resolve that. Sounds risky.

**Comment by zerkms** [comment]
@davidhiendl a few comments above I suggested something that would provide a guaranteed yet simple solution for the problem, for the cost of adding another concept of a `fencer`: https://github.com/rook/rook/issues/1507#issuecomment-457466226

**Comment by lorenz** [comment]
Honestly this really needs to be done by Ceph. I've seen many distributed systems go down or destroy data because of separation of logic and functionality critical to their correct operation (like fencing or locking). And depending on how you attach the RBD device Ceph will absolutely allow multiple R/W attachments.

**Comment by dimm0** [comment]
I've broken several RBDs already like that... With data loss, xfs couldn't recover
https://github.com/rook/rook/issues/3298

**Comment by dimm0** [comment]
Not stale

**Comment by danielxolsson** [comment]
> Solved this by using following config:
> 
> ```
>      terminationGracePeriodSeconds: 0
>      tolerations:
>      - effect: NoExecute
>        key: node.kubernetes.io/unreachable
>        operator: Exists
>        tolerationSeconds: 300
> ```
> 
> Tested with kubernetes v1.12 with rook-ceph v14.2. Note that FlexVolume is used.

I have to inform you that this did not work if the crashed node comes up again. Sometimes the old pod starts and get the volume mounted (which is already mounted) which can cause corrupted filesystem. It happend to me :( Ceph do allow a second mapping (rbd map ...) of volume. However rook-ceph kubernetes driver should not allow multiple rw mount of same volume, but sometimes it does which can cause corrupted filesystem.

**Comment by Antiarchitect** [comment]
Having the same issue with rook 1.1.2 and ceph 14.2.4.
After node outage I have stalled watchers on my rbd images.

**Comment by salanki** [comment]
What's the latest on this? Is any solution actively being worked on? I see it removed from any milestones.

**Comment by travisn** [comment]
This isn't actively being worked on for the flex driver since the focus has turned to the csi driver. Can you confirm if you've seen the same behavior with the csi driver?

@humblec @ShyamsundarR @dillaman Can you take a look at this issue? To summarize, how does the Ceph-CSI driver handle fencing? Does it use its own mechanism or query rbd directly if it is safe to attach a read-write-once volume?

**Comment by salanki** [comment]
I hit the same with CSI today. Node died and Pod refused to start on new node as PVC was still bound to the dead node.

**Comment by gotiniens** [comment]
I can confirm this also happens with the csi driver

**Comment by lorenz** [comment]
Please have a look at my comments further up the thread, it's known why it's happening, it doesn't matter what driver you're using and it's not fixed currently. I've laid out some possible solutions, but I've personally moved away from rook to my own solution and nobody seems to work on it. More "Me Toos" don't really help though.

**Comment by dimm0** [comment]
CSI is now doing `rbd status ...` and checking if volume is already mounted (which is great)
The only part left is to remove the check on kubernetes side I guess..

**Comment by Emil-G** [comment]
So no chance this will be fixed for FlexVolume?  @travisn

**Comment by ShyamsundarR** [comment]
> This isn't actively being worked on for the flex driver since the focus has turned to the csi driver. Can you confirm if you've seen the same behavior with the csi driver?
> 
> @humblec @ShyamsundarR @dillaman Can you take a look at this issue? To summarize, how does the Ceph-CSI driver handle fencing? Does it use its own mechanism or query rbd directly if it is safe to attach a read-write-once volume?

Ceph-CSI also has this issue, as is discussed further in the comments above. There are various issues and active work/thoughts on getting this fixed as below, but till a variety of issues are resolved this needs admin intervention.

[1] ceph-CSI tracker for the issue: https://github.com/ceph/ceph-csi/issues/740
[2] Kubernetes storage SIG thread: https://groups.google.com/forum/#!topic/kubernetes-sig-storage/CRTVASR1atk
[3] Storage SIG presentation as of kubecon Nov, 3rd week, 2019:  https://docs.google.com/presentation/d/1UmZA37nFnp5HxTDtsDgRh0TRbcwtUMzc1XScf5C9Tqc/edit?usp=sharing
[4] Further fencing discussion in ceph-CSI: https://github.com/ceph/ceph-csi/issues/578

**Comment by rofra** [comment]
I'm experiencing the same issue, any update for flex volume ?

```
Events:
  Type     Reason       Age                From                                    Message
  ----     ------       ----               ----                                    -------
  Normal   Scheduled    <unknown>          default-scheduler                       Successfully assigned prod-eu-eth1aaaamainnet/eth1-aaaa-mainnet-0 to node03.cluster03.opchain.link
  Warning  FailedMount  18s (x8 over 82s)  kubelet, node03.cluster03.opchain.link  MountVolume.SetUp failed for volume "pvc-cc6b776d-0377-4c25-a974-e1619b75379b" : mount command failed, status: Failure, reason: Rook: Mount volume failed: failed to attach volume pvc-cc6b776d-0377-4c25-a974-e1619b75379b for pod prod-eu-eth1aaaamainnet/eth1-aaaa-mainnet-0. Volume is already attached by pod prod-eu-eth1aaaamainnet/eth1-aaaa-mainnet-0. Status Pending

```

**Comment by dimm0** [comment]
Delete the node from the cluster is the only way I know for now

**Comment by R-omk** [comment]
it can be used.  Read the sources for more explanations.
https://github.com/kvaps/kube-fencing

Fencing is the only way to guarantee that the volume will not be damaged

**Comment by rofra** [comment]
As far as I understand it, it is due to an error on **StatefulSet** pod deletion, due to "kubectl drain" or classic "kubectl delete pods" commands with **--force --grace-period=0** options.

On https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/pod-safety.md
> Deleting a pod with grace period 0 is called force deletion and will update the pod with a deletionGracePeriodSeconds of 0, and then immediately remove the pod from etcd. Because all communication is asynchronous, force deleting a pod means that the pod processes may continue to run for an arbitrary amount of time. If a higher level component like the StatefulSet controller treats the existence of the pod API object as a strongly consistent entity, deleting the pod in this fashion will violate the at-most-one guarantee we wish to offer for pet sets.

On: https://kubernetes.io/docs/tasks/run-application/force-delete-stateful-set-pod/
> In normal operation of a StatefulSet, there is never a need to force delete a StatefulSet Pod

@R-omk, maybe @dimm0 : Thanks for the kube-fencing recommandation. Otherwise, is there any "shell" or "kubectl' command to run directly on nodes to interrupt the link between a pod and a volume, even  without removing the node from the cluster or usign kube-fencing ?

Maybe a "_docker stop_" command on selected pod name ?

**Comment by R-omk** [comment]
@rofra 
> Maybe a "docker stop" command on selected pod name ?

This is exactly what kubelet is trying to do.

But the most important thing is to unmount the partition, or physically break the link so that nothing could damage the data, this is exactly what fencing does.

So, if you kill processes that could write to the file system on such a partition (it about `docker stop`), dont forget to make sure that file system services (such as journaling) do not modify data either.

In general, it is difficult.
Much easier to just 'unplug the power plug'.

But if you have a data backup and are not afraid to get a broken file system in production, you can take a chance. Most likely nothing bad will happen, but there is no zero chance that something will go wrong.

In my example with fencing, you can implement your custom driver that will work in manual mode. That is, you receive a letter: "your server is down, click **this button** if you are sure that it has finally absolutely down" ....   and... fencing agent just run kubelet delete --force --grace-period=0  


*Please correct me if I am mistaken in something*

**Comment by R-omk** [comment]
Another idea is to have a hot standby pod that will start on volum's snapshot.

Something like operator determines that the main pod is down then 
- removes the old pod from the balancer (ie patch service)
- creates a ceph volume snapshot (I hope is safe for data)
- creates PV linked with the new volume 
- creates a new pod with the new PV

software continues to work while the old pod is stuck waiting for manual intervention

**Comment by Antiarchitect** [comment]
I'm not an expert, but it seems fsfreeze would be nice for data integrity. I saw it appliance in several VMs like VmWare and Proxmox before doing snapshots.

**Comment by dimm0** [comment]
Ceph RBD already has exclusive-lock feature, isn't it supposed to fix this?

**Comment by R-omk** [comment]
@dimm0 I do not quite understand what you mean, but title of this issue contains "can't mount the same volume", so it need to mount another 😄

**Comment by dimm0** [comment]
The issue is that we want to mount the same volume from a different node, right? And in my understanding exclusive-lock should somehow tell ceph which node is now in control of RBD

**Comment by ShyamsundarR** [comment]
As various updates to this issue poured in today, I was also working to establish what problems were present in ceph-csi, that disallowed the mount on a second node. So here is my update, which also closely mirrors other comments made above.

**NOTE:** All testing that I did was using kubernetes v1.17

There are 2 categories of problems being discussed here,
1) Recovering from a node failure: allowing rescheduled pods, from the failed node to other nodes, access to their PVs

2) Fencing/blacklisting stale clients: to prevent inadvertent volume corruption

**Recovering from a node failure: allowing rescheduled pods, from the failed node to other nodes, access to their PVs**

The problems reported in this issue is mostly around, pod never recovering if a node hosting the pod lost communication with the kubernetes api server (or masters). Also, these are for RWO PVs only, and NOT for RWX/ROX access mode PVs (as these can be attached to multiple nodes at the same time).

For pods that are NOT not governed by the StatefulSets controller, the pods recover in ~11+ minutes (based on default settings) due to the following kubernetes events,
  - By default all pods are granted [toleration of 300 seconds](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/#taint-based-evictions) before termination, when the node they are running on enters the "not-ready" or "unreachable" taints
  - **a.** When a node loses connectivity to kubernetes API server, at the end of 300 seconds (5 minutes), the taint_manager marks the pods for deletion due to the above toleration
  - **b.** Based on the controller in use (ReplicaSet or Deployment (but **NOT** StatefulSets)), a new instance of the pod is scheduled to run on a different node
  - At this time, the kubernetes attach-detach controller kicks in, but will not force detach the PV already attached to the unrechable node
  - **c.** The force detach timeout is 6 minutes (governed by maxWaitForUnmountDuration), at which time the PV is force detached and attached to the new node
  - The pod that was created in (b) now gets that PV attached due to (c) and proceeds to move into the running state

During (c) is when CSI plugins even come into play, or are sent the required ControllerPublish/NodeStage calls, before this time duration (IOW, 11+ minutes) there is no way for CSI plugins to react to the environment.

Workarounds:
- Reducing the NoExecute taints, for the pods, to lower values can shorten the 5 minute waits to start the pod on a new node, and thus reduce the overall 11+ minutes wait time for the pod to restart.
- There is no current way to configure a lower value for the maxWaitForUnmountDuration, or avoid the forced detach wait (IOW, the latter 6 minutes)
  - Also, this is a **force detach**, where the status of the failed node is unknown, and it could still be running! This leads to problem (2)

**NOTE:** For StatefulSets, kubernetes waits for termination of the pod on the unreachable node, and only then restarts a new instance on one of the surviving nodes. This results in step (c) above never getting triggered. To overcome this, these StatefulSet pods need to be force deleted or the node needs to be removed from the kubernetes cluster. Post this force-deletion/node-removal, there would still be a 6 minute wait for the force detach to trigger and the new instance to start running.

**Fencing/blacklisting stale clients: to prevent inadvertent volume corruption**

Even with the above 11+ minute wait, if the node that was unreachable, comes back online there is a potential for data corruption in case the volume is written to prior to garbage collection on the pod/mount/image-mapping on the node.

This situation needs fencing, or ceph blacklisting, for RWO mode PVs.

This latter part is being discussed on how to best overcome it and provide the required blacklisting in ceph/ceph-csi#578

**Comment by leseb** [comment]
@ShyamsundarR do you mind sending a doc PR with this write-up? Thanks.

**Comment by ShyamsundarR** [comment]
> @ShyamsundarR do you mind sending a doc PR with this write-up? Thanks.

Will do. I backed up a bit on other activities, so may get the initial PR in late next week or early the week after.

**Comment by abh** [comment]
How do you see which node the system thinks the pod used to run on? I ran into this for my Postgres statefulsets (only, everything else recovered fine...).

The pods don't start with an error along the lines of

```
  Normal   Scheduled    <unknown>         default-scheduler  Successfully assigned postgres/harbor-db-0 to kw1
  Warning  FailedMount  5s (x6 over 22s)  kubelet, kw1       MountVolume.SetUp failed for volume "pvc-84f1a1af-c7be-4b11-ae68-d14984629ad0" : mount command failed, status: Failure, reason: Rook: Mount volume failed: failed to attach volume pvc-84f1a1af-c7be-4b11-ae68-d14984629ad0 for pod postgres/harbor-db-0. Volume is already attached by pod postgres/harbor-db-0. Status Pending
```

... of course the pod that's not starting in this case is `postgres/harbor-db-0`, so I don't get where kubernetes thinks the pod might be running.  We're still using flex volumes (not understanding how to safely upgrade to CSI yet...)

**Comment by papdaniel** [comment]
@abh Workaround to find it:
kubectl -n rook-ceph-system get volumes.rook.io <pvc/pv-name> -o yaml
```
...
apiVersion: rook.io/v1alpha2
attachments:
- clusterName: rook-ceph
  mountDir: /var/lib/kubelet/pods/b019b4d9-f4a3-41a7-a191-2a06c3f6bcf3/volumes/ceph.rook.io~rook-ceph-system/pvc-650bfc29-36a5-11e9-a06c-00505689b514
  node: node01
  podName: gitaly-0
  podNamespace: gitlab
  readOnly: false
kind: Volume
...
```
In `mountDir: /var/lib/kubelet/pods/<pod-uid>/...`  there is your old pod-uid and in `node` its node too.

What I did (I'm not sure what happens in the background so, just careful, but I prayed :) and it worked with gitaly and a postgres cluster):
Changed the old pod-uid to the new pending pod-uid, and `node` to the pending pod node name.
I needed to delete the pod and it was scheduled to another node LoL. So I went back and modified the volumes.rook.io resource,  and again deleted the pod. Played it while the new pod was rescheduled on the same node, what I wrote in `node`.

**Comment by ibotty** [comment]
It should always be possible, albeit possibly dangerous, to delete the `volumeattachments.storage.k8s.io` that belong to the failed node.

I will install a watcher to detect a not responsible node (NotReady status and not pingable for a certain amount of time) that will delete all volumeattachments on my rook clusters.  I cannot tolerate the pet container not to restart until admins take care of something.

**Comment by yanchicago** [comment]
@ibotty Could you help with more details on "a certain amount of time"? If you can't ping the node, what "delete all volumeattachments" involves( i.e., blacklist the volume watcher) ?

**Comment by ibotty** [comment]
If you delete a volumeattachment, kubernetes allows mounting the volume on a different node.

The stopgap measure for me will be a simple loop that watches for node events and acts if the node becomes NotReady.  It will then try to ping the node for some time and will delete all volumeattachments for that node when the node is not responsive to ping either.

That will be dangerous if the node only had network issues and comes back later (not by reboot, but with mounts still intact).

**Comment by jamesorlakin** [comment]
Would it be possible/sensible to also issue a `ceph osd blacklist add <ip>` in that scenario too?

**Comment by ibotty** [comment]
Yes!  That sounds reasonable.  I did not think about that.  It's important to unblacklist the node when it comes back though.  That would make it safer I guess.

@travisn and @leseb, does it make sense to have that in ceph-rook?  Is it safe enough?  I can't think of something better until the right approach (see above) is implemented.

**Comment by yanchicago** [comment]
@ibotty Could you detail on what "delete all volumeattchment" involves?

**Comment by ibotty** [comment]
VolumeAttachment is a kubernetes object type.  So

> kubectl delete volumeattachments.storage.k8s.io csi-[...]

will delete the volumeattachment.

**Comment by zerkms** [comment]
@ibotty the node can become `NotReady` when its kubelet has died, on kubelet port only was firewalled. While the machine was live and with all other processes running just fine.

**Comment by ibotty** [comment]
I know, that's why I said to also ping the node for some amount of time.

**Comment by yanchicago** [comment]
The same issue: https://github.com/rook/rook/issues/4371

**Comment by schaze** [comment]
@ShyamsundarR :
I am having problems with these situations since 1.17.0 as well since some of my hosts sometimes freeze up (no idea why yet)
However for me the pods never come up again on another node as they are stuck waiting for the volume forever. (using a Deployment and rook/ceph block storage volume RWO)

> * **c.** The force detach timeout is 6 minutes (governed by maxWaitForUnmountDuration), at which time the PV is force detached and attached to the new node

--> do I need to add any special configuration for this kind of behaviour (point c) to actually happen?

**Comment by jamesharr** [comment]
I'm seeing a similar situation, luckily in a lab environment. I simulate a node-failure by powering down a physical node, the control-plane attempts to kill the pod and start a new one using the existing PVC. Both sides of the process get stuck. The Kill gets stuck because it can't talk to the node, and the start gets stuck because the volume is in use.

My team would be interested in a manual recovery process while a more automatic process is being worked on.

I know this is a tricky situation with no single "right" answer, but I have a few thoughts:
* If we can't have an automatic decision made, having a documented practice for recovering is worth-while. I might make the first step "Ensure the failed node is actually offline".
* It might be valuable for Workload/PVCs to have some sort of input as the behavior and timeout.
* If Ceph doesn't have a good built-in mechanism for fencing off a failed node after a quorum is established, a novel approach might be to clone/fork a volume into a new volume and use that for the new workload. I've heard this referred to as a crash-consistent snapshot. I'm not sure if K8s architecture facilitates this, but it's something I thought of that might be an option depending on the app.

**Comment by RyuunoAelia** [comment]
The "manual recovery" is pretty simple you can force the k8s API to "forget" the pod. If you are really sure the machine is completely down and it won't corrupt the Data, you can use `kubectl delete pod $pod_name --force --grace-period=0` this will give you a nice warning but the pod will disappear from the API without performing any real action on the node.

**Comment by nixpanic** [comment]
There is a proposal to include node/volume fencing to the CSI specification. Kubernetes (and Ceph-CSI) would be able to initiate a more reliable fail-over when a node becomes unresponsive.

container-storage-interface/spec#477

**Comment by travisn** [comment]
While the [node loss documentation](https://rook.github.io/docs/rook/latest/ceph-csi-troubleshooting.html#node-loss) is help for visibility and attempt to recover, reopening since this is still fundamentally an issue.

**Comment by onedr0p** [comment]
I experienced this issue as well. I am using a `Deployment` with a strategy of type `Recreate` and there's a ceph-block volume attached. My node died overnight and there was a pod stuck in `Terminating` and a new one stuck in `ContainerCreating`.

I was recalling my time using Longhorn and in their documentation they have a specific section on how to handle this issue [here](https://longhorn.io/docs/1.2.4/references/settings/#pod-deletion-policy-when-node-is-down), would it be possible for rook-ceph to provide similar functionality?

**Comment by travisn** [comment]
By itself the policy of force deleting pods is dangerous, which is why we have not yet implemented it. The volume could become corrupted in the following scenario:
1. A node goes offline (e.g. power or partial network outage)
2. Rook force deletes the app pod(s) with rbd volumes on the unresponsive node
3. The app pods start on another node and start writing to the volume
4. The original dead node comes back online, the app pod on that node is still running and writing data to the volume, and now suddenly there are two writers to the volume, which will corrupt the volume.

I don't see any solution yet for the [proposed csi spec update](https://github.com/container-storage-interface/spec/pull/477) that would help here. Thus, we have documented how to [blocklist a dead node](https://rook.github.io/docs/rook/latest/ceph-csi-troubleshooting.html#node-loss) so the pod can then be force deleted. Of course, manual intervention is a big pain point. 

In the meantime, what about the following automation for Rook to help in this scenario? 
1. A node goes offline
2. Rook detects the unresponsive node
3. Wait for some configurable timeout, perhaps with a default of 15 minutes
4. Rook [blocklists the node IP](https://rook.github.io/docs/rook/latest/ceph-csi-troubleshooting.html#node-loss) so Ceph will block all future writes from that node's IP
5. Rook force deletes the application pod(s) on that node, allowing the pod to resume elsewhere

The admin would need to manually unblocklist a node after it is confirmed permanently offline and no longer a threat to data corruption. Rook simply cannot know when a node is safe to remove from the list.

The policy could be controlled by several policy settings such as:
```yaml
nodePolicy:
  offlineWaitTimeBeforeForceDelete: 15m # time to wait before triggering the policy
  maxNodesToBlock: 3 # Stop blocking nodes after this count, to avoid blocking too many nodes. At some point, admin intervention is needed.
  forceDeleteDeployments: true
  forceDeleteStatefulSets: true
  includedNamespaces: # If empty, process app pods in all namespaces, otherwise only this list of namespaces
  - app1
  - app2
  skippedNamepaces: # if specified, process app pods in all namespaces except these
  - app3
  - app4
```


@leseb @BlaineEXE @Madhu-1 @humblec @nixpanic Is this feasible?

**Comment by zerkms** [comment]
I'm not in the list @travisn asked for feedback, but I still have something to say! :-)

What if the step 4 (blocklisting) was implementing through another CRD (eg `RookCephBlockedNode`) with name matching the blocklisted node.

This would have at least these 2 huge benefits:

1. No need for additional tools for monitoring (just a "normal" configured prometheus should already be able to do it)
2. Unblocking would simply be a `kubeclt delete`.

**Comment by ibotty** [comment]
Can't the `ceph-csi`-Pods detect that a node does not have any old mounts left, so it can remove the blocklist entry automatically or instruct some other process to do that?

**Comment by nixpanic** [comment]
> 5. Rook force deletes the application pod(s) on that node, allowing the pod to resume elsewhere

There is still an issue with this step, I think:

_Kubernetes might still assume the volume is mounted on the non-responsive node, and will likely not mount it on an other node (RWO and RWOP volumes). Deleting or updating the `VolumeAttachments` for the unresponsive node might be required?_

In my opinion, blocklisting nodes is a cluster operation, so it is something that Rook should manage. Ceph-CSI only provides a (simple) storage interface for creating/deleting/mounting/unmounting volumes, and is not aware of any state of the Ceph cluster. So a `RookCephBlockedNode` CRD makes sense to me.

**Any automation around this, is dangerous.** It is possible a Pod is running successfully, even when the Kubernetes management plane is non-responsive on a node. Blocklisting the node will then cause the successful running Pod to get I/O errors, and potentially corrupt data... Restarting the Pod on an other node might not be able to recover the corrupted data (a filesystem could require `fsck`, possibly needing user-input, which is not possible, causing a full outage of the Pod, similar for partially written application data).

**Comment by nixpanic** [comment]
Also note that we have implemented [`NetworkFence` from CSI-Addons](https://github.com/csi-addons/spec/blob/main/fence/README.md) in Ceph-CSI. The work is related, maybe Rook can build on top of that.

/cc @Yuggupta27

**Comment by R-omk** [comment]
Here's (https://github.com/kvaps/kube-fencing) something on a similar topic, take a look.

**Comment by zerkms** [comment]
@nixpanic 

> Blocklisting the node will then cause the successful running Pod to get I/O errors, and potentially corrupt data

Shouldn't ceph tolerate network or hardware failures though?

**Comment by leseb** [comment]
While the proposal seems reasonable, I'm afraid this might go beyond what Rook is capable of doing. This looks to me like a machine-level type of problem and Rook is really high in the stack when trying to find out about the issue. Relying on the control plane to determine whether a node is alive or not might be sufficient. As far as I can tell (quick glance) this is how https://github.com/kvaps/kube-fencing seems to be implemented. It's probably better than nothing to rely on the control plane but it's not a 100% source of truth. We need something like Pacemaker along with STONITH (Fencing) agent to handle these cases properly.

**Comment by Madhu-1** [comment]
This is getting addressed at the CSI level as @travisn mentioned in one of the comments (its really going slow) and also as @nixpanic mentioned we already have a CRD to fence and unfence the block of IP addresses using the [csi-addons](https://github.com/csi-addons/spec/blob/main/fence/README.md) repo.

**Comment by nixpanic** [comment]
integrating blocklisting with kube-fencing, or https://github.com/medik8s would surely be a nice approach!

**Comment by nixpanic** [comment]
integrating blocklisting with kube-fencing, or https://github.com/medik8s would surely be a nice approach!

> @nixpanic
> 
> > Blocklisting the node will then cause the successful running Pod to get I/O errors, and potentially corrupt data
> 
> Shouldn't ceph tolerate network or hardware failures though?

Yes, Ceph will take care of OSD issues, the storage managed by Ceph will not have any problems. However there is nothing that Ceph can do about partial data storage by applications (or a filesystem) running on a node that suddenly goes offline. When the application starts (or the filesystem gets mounted) on an other node, the application (or filesystem) might need some recovery (rollback to some checkpoint, or `fsck`).

Imagine a failure that happens while uploading a large file to some application Pod using ext4 on RBD. If the application gets interrupted (node offline) while the upload is in progress, the large fail will be partially stored on the ext4 filesystem backed by Ceph RBD. From a Ceph point of view, a partial file is perfectly acceptable. From the applications point of view, that will need to consume the file, a partial file will not be very useful. Some form of recovery is needed by the application (delete the partial file, and have it re-uploaded). Similar problems can happen when the ext4 filesystem did not sync the latest data to the RBD image, in that case an `fsck` is required. Both types of recovery potentially require some user interaction (request a re-upload, handle `fsck` prompting). Asking users to do something (or informing an application about data issues) is not possible at this time.

**Comment by zerkms** [comment]
> there is nothing that Ceph can do about partial data storage by applications (or a filesystem) running on a node that suddenly goes offline

@nixpanic then ceph/rook should not care about it: hardware is not reliable, network is not reliable, OS and system libs are not reliable, OOM may kill a process at any moment.

> Imagine a failure that happens while uploading a large file to some application Pod using ext4 on RBD. If the application gets interrupted (node offline) while the upload is in progress, the large file will be partially stored on the ext4 filesystem backed by Ceph RBD.

exactly the same will happen if a process is killed by the OOM or of the host machine suddenly losts power.

> Some form of recovery is needed by the application

It must be there whether the app runs on kubernetes with rook/ceph, or on bare metal - the application either handles it or not, it's not ceph/rook's problem/domain of responsibility.

> Similar problems can happen when the ext4 filesystem did not sync the latest data to the RBD image, in that case an fsck is required

The same point as above.

> Asking users to do something (or informing an application about data issues) is not possible at this time.

The same point as above: if the application/fs corrupts during power loss - then it does not matter whether Rook automated it or whether it happened accidentally: it's better to have the application running and then later fix a broken machine, other than have downtime, but with "guaranteed" data integrity (actually not).

**Comment by nixpanic** [comment]
Indeed, by blocklisting the node/vm that does not respond anymore, it should be possible to recover from a certain set of failures. It should just be clear that blocklisting/fencing is not a bulletproof solution. Even when Rook and Ceph-CSI do support the functionality, there will be failures that can not (easily) be recovered.

I hope Rook with Ceph-CSI can consume the [CSI-Addons `NetworkFence`](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md) CRD. Ceph-CSI offers this option at the moment, and more consumers of the API would be most welcome.

**Comment by humblec** [comment]
I agree with @nixpanic , its better to stick to the same fencing model we already have from csi-addons repository.

**Comment by travisn** [comment]
Thanks all for the feedback, what I'm gathering now is that the process would be more like the following:
1. A node goes offline
2. Rook detects the unresponsive node
3. Wait for some configurable timeout, perhaps with a default of 15 minutes
4. Rook creates a NetworkFence CR that fences the unresponsive node (there is no longer a need for the `ceph osd blocklist` command, right?)
5. Rook [deletes the volumeattachments](https://github.com/ceph/ceph-csi/issues/740#issuecomment-770750182) for the rbd PVs on that node
6. Rook force deletes the application pod(s) on that node with rbd volumes, allowing the pod(s) to resume elsewhere

From all the discussion, this will at least allow the application to attempt starting elsewhere to recover automatically in many scenarios. Corruption at the application level will not always be avoided (similar to a power outage), but data will still maintain consistency at the ceph data layer.

**Comment by nixpanic** [comment]
> Rook creates a NetworkFence CR that fences the unresponsive node (there is no longer a need for the `ceph osd blocklist` command, right?)

Indeed, the `NetworkFence` CR will go through the csi-addons controller which passes the request on to the CSI driver. In case of Ceph-CSI, it will execute `ceph osd blocklist` (or similar through go-ceph). @Yuggupta27 is the authority on it, so he may want to correct/confirm it.

**Comment by Yuggupta27** [comment]
> > Rook creates a NetworkFence CR that fences the unresponsive node (there is no longer a need for the `ceph osd blocklist` command, right?)
> 
> Indeed, the `NetworkFence` CR will go through the csi-addons controller which passes the request on to the CSI driver. In case of Ceph-CSI, it will execute `ceph osd blocklist` (or similar through go-ceph). @Yuggupta27 is the authority on it, so he may want to correct/confirm it.

Yes @nixpanic , the NetworkFence CR will call for `ceph osd blocklist add/rm` commands on the underlying ceph cluster, to blocklist/unblock operations from a particular CIDR range(of Node/s).

**Comment by Rakshith-R** [comment]
> Thanks all for the feedback, what I'm gathering now is that the process would be more like the following:
> 
> 1. A node goes offline
> 2. Rook detects the unresponsive node
> 3. Wait for some configurable timeout, perhaps with a default of 15 minutes
> 4. Rook creates a NetworkFence CR that fences the unresponsive node (there is no longer a need for the `ceph osd blocklist` command, right?)
> 5. Rook [deletes the volumeattachments](https://github.com/ceph/ceph-csi/issues/740#issuecomment-770750182) for the rbd PVs on that node
> 6. Rook force deletes the application pod(s) on that node with rbd volumes, allowing the pod(s) to resume elsewhere
> 
> From all the discussion, this will at least allow the application to attempt starting elsewhere to recover automatically in many scenarios. Corruption at the application level will not always be avoided (similar to a power outage), but data will still maintain consistency at the ceph data layer.


Referring to this upstream proposal,
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2268-non-graceful-shutdown#proposal

 I think we need to let user handle step 1-3 and apply `out-of-service` taint.
 k8s with this new feature will take care of step 5 & 6.
 Rook can take care of step 4 as follows:

- Rook can detect `out-of-service` taint on node.

- Create NetworkFence CR to blocklist/Fence the node IP.

**Comment by travisn** [comment]
Agreed that it is best to build on this KEP for detecting when nodes are offline and allowing K8s to take care of force deleting the app pods on the failed node. Then the remaining work is simpler for Rook to only worry about creating the NetworkFence CR.

In reality it will take quite some time to get the changes from the KEP into K8s. There is also not yet a proposal to automatically mark a node offline. Thus, I would still like to explore an option for Rook to take this step of automatically fencing nodes when they appear to be offline. I don't see it ever being enabled by default, but will be a very useful option for admins to have. Somehow I don't anticipate we will ever have a full solution from the K8s for this issue since it's more of a storage problem for rbd to deal with to avoid corruption.

**Comment by Rakshith-R** [comment]
> Agreed that it is best to build on this KEP for detecting when nodes are offline and allowing K8s to take care of force deleting the app pods on the failed node. Then the remaining work is simpler for Rook to only worry about creating the NetworkFence CR.

:+1:  Sounds good to me.

> In reality it will take quite some time to get the changes from the KEP into K8s. There is also not yet a proposal to automatically mark a node offline. Thus, I would still like to explore an option for Rook to take this step of automatically fencing nodes when they appear to be offline. I don't see it ever being enabled by default, but will be a very useful option for admins to have. Somehow I don't anticipate we will ever have a full solution from the K8s for this issue since it's more of a storage problem for rbd to deal with to avoid corruption.

We can open a new issue for automating this behaviour and continue discussion on conditions for node loss and its detection over there if that works with everyone.

**Comment by subhamkrai** [comment]
design doc in progress https://github.com/rook/rook/pull/12253 here
